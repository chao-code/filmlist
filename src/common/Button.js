import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Link } from 'react-router-dom'

const Button = props => {
  const { children,
    className,
    href,
    primary,
    secondary,
    danger,
    link,
    small,
    ...attrs
  } = props
  const bsClass = classNames('btn', className, {
    'btn-primary': primary,
    'btn-outline-secondary': secondary,
    'btn-danger': danger,
    'btn-link': link,
    'btn-sm': small,
  })

  return (
    href ? (
      <Link className={bsClass} to={href} {...attrs}>{children}</Link>
    ) : (
      <button className={bsClass} {...attrs}>{children}</button>
    )
  )
}

Button.propTypes = {
  children: PropTypes.string,
  className: PropTypes.string,
  href: PropTypes.string,
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  danger: PropTypes.bool,
  link: PropTypes.bool,
  small: PropTypes.bool,
}

Button.defaultProps = {
  children: ' ',
  className: '',
  href: '',
  primary: false,
  secondary: false,
  danger: false,
  link: false,
  small: false,
}

export default Button
