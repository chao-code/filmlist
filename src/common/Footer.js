const Footer = () => (
  <footer className="container">
    <hr />
    <p className="text-center my-3">
      <small>Â© 2018. Powered by <span role="img" aria-label="Poop">ğŸ’©ğŸ’©</span></small>
    </p>
  </footer>
)

export default Footer
