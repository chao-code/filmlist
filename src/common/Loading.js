const Loading = () => (
  <div className="d-flex justify-content-center my-5">
    <div className="loading" />
  </div>
)

export default Loading
