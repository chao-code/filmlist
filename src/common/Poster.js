import PropTypes from 'prop-types'
import cropPoster from '../utils/cropPoster'

const Poster = ({ poster, crop }) => {
  const bsClass = 'img-fluid card'
  const src = crop ? cropPoster(poster) : poster


  return (
    <img className={bsClass} src={src} alt="Poster" />
  )
}

Poster.propTypes = {
  poster: PropTypes.string,
  crop: PropTypes.bool,
}

Poster.defaultProps = {
  poster: '',
  crop: false,
}

export default Poster
