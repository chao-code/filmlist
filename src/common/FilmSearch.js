import { Component, Fragment } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { loadSearch } from '../redux/actions'
import debounce from '../utils/debounce'
import Poster from './Poster'
import TitleYear from './TitleYear'

class FilmSearch extends Component {
  static propTypes = {
    inputClass: PropTypes.string,
    resultsContainerClass: PropTypes.string,
    placeholder: PropTypes.string,
    onResultClick: PropTypes.func,
    keepFocusOnResultClick: PropTypes.bool,
    loadSearch: PropTypes.func.isRequired,
    searchResultsByQuery: PropTypes.objectOf(PropTypes.shape({
      isFetching: PropTypes.bool,
      results: PropTypes.array,
    })).isRequired,
  }

  static defaultProps = {
    inputClass: '',
    resultsContainerClass: '',
    placeholder: '',
    onResultClick: () => {},
    keepFocusOnResultClick: false,
  }

  state = {
    value: '',
    results: null,
    isLoading: false,
    isFocused: false,
  }

  onInputChange = event => {
    const value = event.target.value
    this.setState({ value })
    this.search(value.trim())
  }

  onInputFocus = () => {
    this.setState({ isFocused: true })
  }

  onInputBlur = () => {
    if (!this.justMouseDownOnResultsContainer) {
      this.setState({ isFocused: false })
    }
  }

  onResultClick = result => {
    this.justClickedOnResult = true
    this.props.onResultClick(result)
    this.setState({
      value: '',
      results: null,
      isFocused: false,
    })
  }

  onResultsContainerMouseDown = () => {
    this.justMouseDownOnResultsContainer = true
  }

  onResultsContainerClick = () => {
    this.justMouseDownOnResultsContainer = false
    if (this.justClickedOnResult) {
      this.justClickedOnResult = false
      this.props.keepFocusOnResultClick && this.input.focus()
    } else {
      this.input.focus()
    }
  }

  search = debounce(query => {
    if (query === '') {
      this.setState({ results: null })
    } else {
      this.setState({ isLoading: true })
      this.props.loadSearch(query)
      .then(() => this.setState({
        results: this.props.searchResultsByQuery[query].results,
        isLoading: false,
      }))
    }
  }, 500)

  renderResult(result) {
    const { poster, title, year, imdbID } = result
    const liProps = {
      className: 'dropdown-item',
      onClick: () => { this.onResultClick(result) },
      key: imdbID,
    }

    return (
      <li {...liProps}>
        <div className="row cursor-pointer">
          <div className="col-3">
            <Poster poster={poster} />
          </div>
          <div className="col">
            <TitleYear title={title} year={year} />
          </div>
        </div>
      </li>
    )
  }

  renderNoResults() {
    return (
      <div className="dropdown-item no-results">
        Sorry, there are no results for that search.
      </div>
    )
  }

  renderResultsContainer() {
    const { resultsContainerClass } = this.props
    const { results } = this.state
    const containerProps = {
      className: classNames('dropdown-menu bd-search-results', resultsContainerClass),
      onMouseDown: this.onResultsContainerMouseDown,
      onClick: this.onResultsContainerClick,
    }

    return (
      <div {...containerProps}>
        {results.length > 0 ? (
          results.map(result => this.renderResult(result))
        ) : (
          this.renderNoResults()
        )}
      </div>
    )
  }

  render() {
    const { inputClass, placeholder } = this.props
    const { value, results, isLoading, isFocused } = this.state
    const inputProps = {
      className: classNames('form-control', { 'bd-search-loading': isLoading }, inputClass),
      value,
      placeholder,
      type: 'search',
      onChange: this.onInputChange,
      onFocus: this.onInputFocus,
      onBlur: this.onInputBlur,
      ref: input => { this.input = input },
    }

    return (
      <Fragment>
        <input {...inputProps} />
        {results && isFocused &&
          this.renderResultsContainer()
        }
      </Fragment>
    )
  }
}

export { FilmSearch as _FilmSearch }

const mapStateToProps = ({ searchResultsByQuery }) => ({ searchResultsByQuery })

const mapDispatchToProps = { loadSearch }

export default connect(mapStateToProps, mapDispatchToProps)(FilmSearch)
