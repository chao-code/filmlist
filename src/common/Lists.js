import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Link } from 'react-router-dom'
import Poster from './Poster'

const renderList = (list, compact) => {
  const { listID, posters, subject, listedBy, filmCount, description } = list
  const url = `/list/${listID}`
  const row = classNames({ row: !compact })
  const col4 = classNames({ 'col-md-4': !compact })
  const col8 = classNames({ 'col-md-8': !compact })
  const text = `${filmCount} ${filmCount === 1 ? 'film' : 'films'} listed by ${listedBy}`
  const offset = description.indexOf('\n')
  const brief = offset < 0 ? description : description.slice(0, offset)

  return (
    <li className="my-3" key={listID}>
      <div className={row}>
        <div className={col4}>
          <div className="row no-gutters mb-2">
            {[0, 1, 2, 3, 4].map(i => (
              <div className="col" key={i}>
                {posters[i] &&
                  <Link to={url}>
                    <Poster poster={posters[i]} crop />
                  </Link>
                }
              </div>
            ))}
          </div>
        </div>
        <div className={col8}>
          {compact ? (
            <h6 className="mb-0">
              <Link to={url} className="link">{subject}</Link>
            </h6>
          ) : (
            <h5 className="mb-0">
              <Link to={url} className="link">{subject}</Link>
            </h5>
          )}
          <small className="text-muted">{text}</small>
          {!compact &&
            <div>{brief}</div>
          }
        </div>
      </div>
    </li>
  )
}

const Lists = ({ lists, compact }) => (
  <ul className="list-unstyled">
    {lists.map(list => renderList(list, compact))}
  </ul>
)

Lists.propTypes = {
  lists: PropTypes.arrayOf(PropTypes.shape({
    listID: PropTypes.string,
    posters: PropTypes.array,
    listedBy: PropTypes.string,
    filmCount: PropTypes.number,
    description: PropTypes.string,
  })).isRequired,
  compact: PropTypes.bool,
}

Lists.defaultProps = {
  compact: false,
}

export default Lists
