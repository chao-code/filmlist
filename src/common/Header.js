import { Link, withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import Button from './Button'
import FilmSearch from './FilmSearch'

const Header = ({ history, location }) => {
  const onResultClick = film => history.push(`/film/${film.imdbID}`)

  const filmSearchProps = {
    inputClass: 'form-control-sm',
    placeholder: 'Search...',
    onResultClick,
  }

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <div className="container">
        <Link className="navbar-brand" to="/">FilmList</Link>
        <button className="navbar-toggler" data-toggle="collapse" data-target="#navbarMenu">
          <span className="navbar-toggler-icon" />
        </button>

        <div className="collapse navbar-collapse justify-content-end" id="navbarMenu">
          <form className="form-inline my-2 mr-sm-3 bd-search">
            <FilmSearch {...filmSearchProps} />
          </form>
          <Button
            className="my-2 my-md-auto"
            primary
            small
            href="/newlist"
            replace={location.pathname.slice(0, 8) === '/newlist'}
          >Create a List</Button>
        </div>
      </div>
    </nav>
  )
}

Header.propTypes = {
  history: PropTypes.shape({ push: PropTypes.func }).isRequired,
  location: PropTypes.shape({ pathname: PropTypes.string }).isRequired,
}

export { Header as _Header }

export default withRouter(Header)
