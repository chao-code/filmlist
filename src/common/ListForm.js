import { Component } from 'react'
import PropTypes from 'prop-types'
import update from 'immutability-helper'
import classNames from 'classnames'
import FilmSearch from './FilmSearch'
import Button from './Button'
import Poster from './Poster'
import TitleYear from './TitleYear'

const indexArray = n => Array(n).fill().map((_, i) => `${i + 1}`)

class ListForm extends Component {
  static propTypes = {
    list: PropTypes.shape({
      subject: PropTypes.string,
      listedBy: PropTypes.string,
      passcode: PropTypes.string,
      description: PropTypes.string,
      films: PropTypes.array,
    }),
    onSave: PropTypes.func.isRequired,
    onCancel: PropTypes.func.isRequired,
  }

  static defaultProps = {
    list: {
      subject: '',
      listedBy: '',
      passcode: '',
      description: '',
      films: [],
    },
  }

  state = {
    ...this.props.list,
    indexInputs: indexArray(this.props.list.films.length),
    showInvalid: false,
    alert: null,
  }

  onInputChange = event => {
    const { value, name } = event.target
    this.setState({
      [name]: value,
      alert: null,
    })
  }

  onIndexInputChange = index => event => {
    const { value } = event.target
    this.setState(prevState => ({
      indexInputs: update(prevState.indexInputs, {
        [index]: { $set: value },
      }),
      alert: Number.isInteger(+value) ? null : {
        type: 'warning',
        message: `"${value}" is not a valid index. Please enter an number then press Enter.`,
      },
    }))
  }

  onCommentInputChange = index => event => {
    const { value } = event.target
    this.setState(prevState => ({
      films: update(prevState.films, {
        [index]: {
          comment: { $set: value },
        },
      }),
    }))
  }

  dismissAlert = () => this.setState({ alert: null })

  addFilm = newFilm => this.setState(prevState => {
    const { films, indexInputs } = prevState
    if (films.find(film => film.imdbID === newFilm.imdbID)) {
      return {
        alert: {
          type: 'info',
          message: `${newFilm.title} (${newFilm.year}) is already in the list.`,
        },
      }
    }
    return {
      films: [...films, newFilm],
      indexInputs: [...indexInputs, `${indexInputs.length + 1}`],
      alert: null,
    }
  })

  removeFilm = index => () => this.setState(prevState => {
    const { films } = prevState
    return {
      films: [...films.slice(0, index), ...films.slice(index + 1)],
      indexInputs: indexArray(films.length - 1),
      alert: null,
    }
  })

  moveFilm = index => event => {
    if (event.key !== 'Enter') return
    const { value } = event.target
    let newIndex = value - 1
    if (newIndex === index) return
    if (Number.isNaN(newIndex)) {
      this.setState(prevState => {
        const { indexInputs, films } = prevState
        const { title, year } = films[index]
        return {
          indexInputs: update(indexInputs, {
            [index]: { $set: `${index + 1}` },
          }),
          alert: {
            type: 'danger',
            message: `Cannot move ${title} (${year}). "${value}" is not a valid index.`,
          },
        }
      })
    } else {
      this.setState(({ films }) => {
        const film = films[index]
        if (newIndex < 0) {
          newIndex = 0
        }
        if (newIndex >= films.length) {
          newIndex = films.length - 1
        }
        return {
          films: update(films, { $splice: [[index, 1], [newIndex, 0, film]] }),
          indexInputs: indexArray(films.length),
          alert: null,
        }
      })
    }
  }

  save = () => {
    const { showInvalid, alert, indexInputs, ...list } = this.state
    if (!list.films.length) {
      this.setState({
        showInvalid: true,
        alert: {
          type: 'danger',
          message: 'Cannot save list. Please add a film.',
        },
      })
    } else if (!list.subject || !list.listedBy) {
      this.setState({
        showInvalid: true,
        alert: {
          type: 'danger',
          message: 'Cannot save list. Please fill in all required fields.',
        },
      })
    } else {
      this.props.onSave(list)
    }
  }

  cancel = () => this.props.onCancel()

  renderListInfo() {
    const { subject, listedBy, passcode, description, showInvalid } = this.state
    return (
      <div className="row">
        <div className="col-md">
          <div className="form-group">
            <label className="form-control-label" htmlFor="subjectInput">Subject</label>
            <input
              className={classNames('form-control', { 'is-invalid': showInvalid && !subject })}
              type="text"
              id="subjectInput"
              name="subject"
              value={subject || ''}
              onChange={this.onInputChange}
            />
            <small className={showInvalid && !subject ? 'invalid-feedback' : 'form-text text-muted'}>
              Required
            </small>
          </div>
          <div className="form-group">
            <label className="form-control-label" htmlFor="listedByInput">Listed By</label>
            <input
              className={classNames('form-control', { 'is-invalid': showInvalid && !listedBy })}
              type="text"
              id="listedByInput"
              name="listedBy"
              value={listedBy || ''}
              onChange={this.onInputChange}
            />
            <small className={showInvalid && !listedBy ? 'invalid-feedback' : 'form-text text-muted'}>
              Required
            </small>
          </div>
          <div className="form-group">
            <label className="form-control-label" htmlFor="passcodeInput">Passcode</label>
            <input
              className="form-control"
              type="text"
              id="passcodeInput"
              name="passcode"
              value={passcode || ''}
              onChange={this.onInputChange}
            />
            <small className="form-text text-muted">
              Use Passcode to edit or delete this list
            </small>
          </div>
        </div>
        <div className="col-md">
          <div className="form-group">
            <label className="form-control-label" htmlFor="descriptionInput">Description</label>
            <textarea
              className="form-control"
              id="descriptionInput"
              rows="10"
              name="description"
              value={description || ''}
              onChange={this.onInputChange}
            />
          </div>
        </div>
      </div>
    )
  }

  renderFormActions() {
    const { showInvalid, films } = this.state
    return (
      <div className="row my-4">
        <div className="col col-md">
          <div className="input-group">
            <div className="input-group-prepend">
              <div className="input-group-text">Add to List</div>
            </div>
            <FilmSearch
              inputClass={classNames({ 'is-invalid': showInvalid && !films.length })}
              resultsContainerClass="film-results"
              placeholder="Enter Film Name..."
              onResultClick={this.addFilm}
            />
          </div>
        </div>
        <div className="col-5 col-md">
          <Button primary className="float-right" onClick={this.save}>Save</Button>
          <Button secondary className="float-right mr-3" onClick={this.cancel}>Cancel</Button>
        </div>
      </div>
    )
  }

  renderAlertMessage() {
    const { type, message } = this.state.alert
    return (
      <div className={`alert fade show alert-${type}`}>
        {message}
        <button type="button" className="close" aria-label="Close" onClick={this.dismissAlert}>
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
    )
  }

  renderListFilms() {
    const { films, indexInputs } = this.state
    return (
      <ul className="list-group">
        {films.map(({ title, year, poster, comment, imdbID }, index) => (
          <li className="list-group-item" key={imdbID}>
            <div className="row px-3">
              <div className="col-2 col-lg-1 px-1 px-md-2">
                <Poster poster={poster} />
              </div>
              <div className="col">
                <input
                  className="form-control text-center mr-2 py-1 px-1 mb-1 order-input"
                  type="text"
                  value={indexInputs[index]}
                  onChange={this.onIndexInputChange(index)}
                  onKeyPress={this.moveFilm(index)}
                />
                <h5 className="film-title">
                  <TitleYear title={title} year={year} />
                </h5>
                <input
                  className="form-control form-control-sm mt-3"
                  type="text"
                  placeholder="Add a Comment"
                  value={comment || ''}
                  onChange={this.onCommentInputChange(index)}
                />
              </div>
              <div className="col-1">
                <button className="close" onClick={this.removeFilm(index)}>
                  <span>&times;</span>
                </button>
              </div>
            </div>
          </li>
        ))}
      </ul>
    )
  }

  render() {
    return (
      <div>
        {this.renderListInfo()}
        {this.renderFormActions()}
        {this.state.alert &&
        this.renderAlertMessage()
        }
        {this.renderListFilms()}
      </div>
    )
  }
}

export default ListForm
