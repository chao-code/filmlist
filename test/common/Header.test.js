import { shallow } from 'enzyme'
import { _Header as Header } from '../../src/common/Header'

const props = {
  history: {
    push: jest.fn(),
  },
  location: {
    pathname: '/',
  },
}

describe('<Header />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Header {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
