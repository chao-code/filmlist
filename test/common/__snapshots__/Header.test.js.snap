// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render correctly 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <div
    className="container"
  >
    <Link
      className="navbar-brand"
      replace={false}
      to="/"
    >
      FilmList
    </Link>
    <button
      className="navbar-toggler"
      data-target="#navbarMenu"
      data-toggle="collapse"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse justify-content-end"
      id="navbarMenu"
    >
      <form
        className="form-inline my-2 mr-sm-3 bd-search"
      >
        <Connect(FilmSearch)
          inputClass="form-control-sm"
          onResultClick={[Function]}
          placeholder="Search..."
        />
      </form>
      <Button
        className="my-2 my-md-auto"
        danger={false}
        href="/newlist"
        link={false}
        primary={true}
        replace={false}
        secondary={false}
        small={true}
      >
        Create a List
      </Button>
    </div>
  </div>
</nav>
`;
